### DISCUSSION TITLE
Does pkl-go need the generator?

### DISCUSSION THREAD

**Discussion by StefMa** [apple/pkl#849] [General]
I’m considering adding pkl to one of our Go projects. What I don’t like so far is that it seems to require installing the code generation tool first and then generating the code 🙃.

However, looking at the go-examples repo, it seems like the code generation tool isn’t strictly required. It might give you some convenient functions, but that’s about it. Am I right in assuming that? 🤔

Where I’m coming from
Let me explain my reasoning here.
Marshaling and unmarshaling other configuration formats (like JSON) also don’t require a code generation tool. You can simply define a struct, call json.Marshal, and include the struct and JSON as text.

Of course, there are tools like "JSON to Go Struct" (websites), but in general, that approach feels simpler to me. Instead of relying on a code generation tool, you can just work with the data directly.

---

So, if pkl-go could follow a similar flow to other configuration libraries, wouldn’t it make more sense to:

1. Provide a website for generating "pkl to Go structs"

2. Include an example that doesn’t rely on the code generation tool

Maybe I’m missing something, and the code generation tool is super handy. But at first glance, it feels like a bit of unnecessary overhead to me.

Thanks in advance 🎄

**Comment by bioball** [comment]
Codegen is optional, and not required to use Pkl with Go. We have a (admittedly small) section about it on our website: https://pkl-lang.org/go/current/evaluation.html#without-code-generation

### Polymorphism

If you are implementing your own codegen and are doing any polymorphism, make sure to read through how pkl-go expects classes to be represented as Go:

https://pkl-lang.org/go/current/codegen.html#classes

Also: be sure to include a `RegisterMapping` call, so that the mapper knows how Pkl classes should turn into Go structs/interfaces, like so: https://github.com/apple/pkl-go/blob/5a09070e08915ff745c2a6b82e3a4f2fc8906739/codegen/snippet-tests/output/bugholder/init.pkl.go

**Comment by knadh** [comment]
Hi @bioball. The [_Without code generation_](https://pkl-lang.org/go/current/evaluation.html#without-code-generation) example still seems to require a `pkl` bin to be present. I wasn't able to find anything in the docs that talks about the bin.

```
$ go run main.go
panic: exec: "pkl": executable file not found in $PATH

goroutine 1 [running]:
main.main()
        /tmp/pkl/main.go:18 +0x1d4
exit status 2
```

**Reply by bioball** [reply]
Yes, that's how pkl-go interacts with the Pkl runtime. pkl-go itself does not do the evaluation; instead, it needs an implementation of Pkl to do actual the evaluation.

Currently, pkl-go can only shell out to a child process. In the future, we hope to also provide a C library that can be linked to.

**Reply by StefMa** [reply]
Hmm, that is sad 😔.
Does both solutions, code generation and no-code generation, needs Pkl on the path? 🤔

We should definitely update the doc with that information.

**Reply by HT154** [reply]
It's documented, though possibly obliquely:

In the Quickstart:
> Also, ensure that Pkl is [installed on your machine](https://pkl-lang.org/main/current/pkl-cli/index.html#installation).

On the Evaluator API page:
> Evaluation occurs by spawning the pkl cli as a child process, and communicates via message passing.

**Reply by StefMa** [reply]
> It's documented, though possibly obliquely:
> 
> In the Quickstart:
> > Also, ensure that Pkl is [installed on your machine](https://pkl-lang.org/main/current/pkl-cli/index.html#installation).
> 

I have the feeling that it should documented in the go quickstart: https://pkl-lang.org/go/current/quickstart.html

Looking at the cli documentation when working with go code is not so obvious 🙃.

Also, I wound expected that if working with such an library, the library brings everything it needs...

**Reply by HT154** [reply]
> > It's documented, though possibly obliquely:
> > 
> > In the Quickstart:
> > > Also, ensure that Pkl is [installed on your machine](https://pkl-lang.org/main/current/pkl-cli/index.html#installation).
> > 
> 
> I have the feeling that it should documented in the go quickstart: https://pkl-lang.org/go/current/quickstart.html

That's where I quoted from! Right at the end of the Install dependencies section.

> Also, I wound expected that if working with such an library, the library brings everything it needs... 

I tend to agree, but unfortunately this does not align with the go language's idea of package management. There's no reasonable way for a library like pkl-go to pull around a dependency like the Pkl CLI.

**Comment by Hipska** [comment]
Not sure if this related to the topic, but is it possible to generate Pkl scheme from Go structs?

**Reply by bioball** [reply]
Indirectly, yes; you can turn Go types into JSON Schema, then from JSON Schema to Pkl.

Go to JSON Schema: https://github.com/invopop/jsonschema

JSON Schema to Pkl: https://pkl-lang.org/package-docs/pkg.pkl-lang.org/pkl-pantry/org.json_schema.contrib/current/generate/index.html

**Reply by Hipska** [reply]
Feels like this could be a good/great addition to the `pkl-go` binding?

**Reply by bioball** [reply]
It would be nice to have, but, this isn't really something that is on our roadmap. Also, I'm not really sure what the path would be to consume data from Go in Pkl if it wasn't through an intermediary like JSON.

What's your use-case for this?

**Reply by Hipska** [reply]
Being able to generate definitions for all telegraf plugins without the need te create them manually.
